hw/mcu/atmel/sam3x8/src/hal_uart.c

```diff
@@ -0,0 +1,306 @@
+        break;
+        case 2:
+            uart->options.stop_bits = US_MR_NBSTOP_2_BIT;
+        break;
+        default:
+            return -1;
+    }   
+
+    /* Ignore flow control, but throw error on illegal argument */
+    switch (flow_ctl){
+        case HAL_UART_FLOW_CTL_NONE:
+        case HAL_UART_FLOW_CTL_RTS_CTS:
+            break;
+        default:
+            return -1;
+    }
+    uart->u_open = 1;
+    /* clock rate should be something like - sysclk_get_peripheral_hz() */
+    if (!usart_init_rs232(uart->uart, &(uart->options), USART_CLOCK_RATE)){
+        return -1;
+    }
+
+    return 0;
+}
+
+/**
+ * hal uart config
+ *
+ * Applies given configuration to UART.
+ */
+int hal_uart_config(int uart, int32_t speed, uint8_t databits, uint8_t stopbits,
+  enum hal_uart_parity parity, enum hal_uart_flow_ctl flow_ctl){
+      void *uart_ptr = uarts[uart].uart;
+      if (is_usart(uart_ptr)){
+          /* USART configuration */
+          /* Assume USART clock and Board are initialized */
+        hal_usart_config(&uarts[uart], speed, databits, stopbits, parity, flow_ctl);
+      }else{
+          /* UART configuration */
+      }
+      return 0;
+}
+
+/*
+ * Close UART port. Can call hal_uart_config() with different settings after
+ * calling this.
+ */
+int hal_uart_close(int port){
+    hal_uart_t *u = &uarts[port];
+    if (is_usart(u->uart)){
+        usart_disable_rx(u->uart);
+        usart_disable_tx(u->uart);
+    }else{
+
+    }
+    u->u_open = 0;
+    return 0;
+}
+
+/**
+ * hal uart start tx
+ *
+ * More data queued for transmission. UART driver will start asking for that
+ * data.
+ */
+void hal_uart_start_tx(int port){
+    hal_uart_t *u = &uarts[port];
+    int sz, i = 0;
+    if (port >= UART_COUNT || !u->u_open){
+        return;
+    }
+    /* While there is data to write */
+    u->tx_on = 1;
+    while ((sz = fill_tx_buffer(u))){
+        if (is_usart(u)){
+            usart_enable_tx(u->uart);
+            for (i = 0; i < sz; i++){
+                /* TODO - replace with a sleep */
+                while (!usart_is_tx_ready(u->uart));
+                usart_write(u->uart, (uint32_t) u->txdata[i]);  
+            }
+            usart_disable_tx(u->uart);
+        }else{
+
+        }
+    }
+
+    /* Call back the user tx done function */
+    if (u->u_tx_done){
+        u->u_tx_done(u->u_func_arg);
+    }
+    u->tx_on = 0;
+}
+
+/**
+ * hal uart start rx
+ *
+ * Upper layers have consumed some data, and are now ready to receive more.
+ * This is meaningful after uart_rx_char callback has returned -1 telling
+ * that no more data can be accepted.
+ */
+void hal_uart_start_rx(int port){
+    hal_uart_t *u = &uarts[port];
+    if (port >= UART_COUNT || !u->u_open){
+        return;
+    }
+    if (is_usart(u->uart)){
+        usart_enable_rx(u->uart);
+        while (!usart_is_rx_buf_end(u->uart)){
+            /* TODO - replace with a sleep */
+            while(!usart_is_rx_ready(u->uart));

```

Comment:
While we could leave this as is for testing, blocking stuff like this could cause problems when the RTOS needs to switch tasks.

Reply:
Yeah, I wasn't sure how to tell it to sleep. Normally I would use sleep(3), but that seems to be linux specific.